### PR TITLE
Add uncertainty quantification methods

### PR DESCRIPTION
**Pull request by neale**
This PR includes 3 changes.

* Add an eval_uncertainty function for hypernetwork and particle_vi algorithm. This function should compute AUC with respect to a defined outlier dataset.
* Refactor utils/datagen.py to be more concise, and support loading of user-specified subsets of classes for MNIST/CIFAR-10.
* Add minmax (L2) method to particle_vi_algorithm and corresponding tests.

The other addition is two gin files for running amortized par_vi and functional par_vi on MNIST.

### CODE REVIEW

File: alf/algorithms/functional_particle_vi_algorithm.py [hunk 1]
```diff
@@ -24,14 +25,21 @@
...
 from alf.data_structures import AlgStep, LossInfo, namedtuple
 from alf.networks import EncodingNetwork, ParamNetwork
 from alf.tensor_specs import TensorSpec
 from alf.nest.utils import get_outer_rank
 from alf.utils import common, math_ops, summary_utils
 from alf.utils.summary_utils import record_time
+try:
+    from sklean.metrics import roc_auc_score
+except:
+    pass
```
Comment (runjerry) [inline alf/algorithms/functional_particle_vi_algorithm.py]: roc_auc_score not used in this file?

File: alf/algorithms/functional_particle_vi_algorithm_test.py [hunk 2]
```diff
@@ -56,10 +57,15 @@ def assertArrayGreater(self, x, y, eps):
...
 
-    @parameterized.parameters(('gfsf'), ('svgd'), ('gfsf', True),
-                              ('svgd', True))
+    @parameterized.parameters(
+        ('svgd', False),
+        ('gfsf', False),
+        ('svgd', True),
+        ('gfsf', True),
+        ('minmax', False),
+    )
```
Comment (runjerry) [inline alf/algorithms/functional_particle_vi_algorithm_test.py:67]: evaluate_uncertainty() has not been tested?

File: alf/algorithms/hypernetwork_algorithm.py [hunk 3]
```diff
@@ -59,6 +90,58 @@ def regression_loss(output, target):
...
+        targets.append(target.view(-1))
+        output, _ = model(data)
+        if output.dim() == 2:
+            output = output.unsqueeze(1)
+        output = output.transpose(0, 1)
+        outputs.append(output)
+    model_outputs = torch.cat(outputs, dim=1)  # [N, B, D]
+    return model_outputs, torch.cat(targets, -1).view(-1)
+
+
```
Comment (runjerry) [inline alf/algorithms/hypernetwork_algorithm.py:37]: auc_score() and predict_dataset() seem to be used by not only hypernetwork_algorithm, maybe can be put into utils? Same for classification_loss and regression_loss?

File: alf/algorithms/hypernetwork_algorithm_test.py [hunk 4]
```diff
@@ -54,13 +55,14 @@ def assertArrayGreater(self, x, y, eps):
         self.assertEqual(x.shape, y.shape)
         self.assertGreater(float(torch.min(x - y)), eps)
 
-    @parameterized.parameters(('gfsf'), ('svgd2'), ('svgd3'), ('minmax'),
+    @parameterized.parameters(('gfsf', False), ('svgd2', False),
+                              ('svgd3', False), ('minmax', False),
```
Comment (runjerry) [inline alf/algorithms/hypernetwork_algorithm_test.py:62]: evaluate_uncertainty() not been tested.

File: alf/utils/datagen.py [hunk 5]
```diff
@@ -11,18 +11,37 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""Adapted from the following:
+"""Set of methods for loading datasets for supervised learning.
+Adapted from the following:
```
Comment (runjerry) [inline alf/utils/datagen.py]: Maybe we do not need this comment given you are a contributor of this file now?

**Review by runjerry — CHANGES_REQUESTED** [review]

File: alf/algorithms/hypernetwork_algorithm.py (hunk 3 above)
Comment (neale) [reply alf/algorithms/hypernetwork_algorithm.py:37]: in the latest commit, I pushed an sl_utils.py file, containing common functions like classification/regression loss, as well as ood utilities.

File: alf/algorithms/hypernetwork_algorithm_test.py (hunk 4 above)
Comment (neale) [reply alf/algorithms/hypernetwork_algorithm_test.py:62]: tested. 
My comment here is that I could just test this function with random inputs, instead of actual MNIST images, since I don't care about "passing the test" in the usual way, I just want eval_uncertainty to run.

File: alf/utils/datagen.py (hunk 5 above)
Comment (neale) [reply alf/utils/datagen.py]: agreed

File: alf/algorithms/hypernetwork_algorithm_test.py (hunk 4 above)
Comment (runjerry) [reply alf/algorithms/hypernetwork_algorithm_test.py:62]: Agreed, just make sure it runs.

Comment (neale) [reply alf/algorithms/hypernetwork_algorithm_test.py:62]: will update

File: alf/algorithms/config.py [hunk 6]
```diff
@@ -122,7 +125,10 @@ def __init__(self,
                 chance of action sampling instead of taking argmax. This can
                 help prevent a dead loop in some deterministic environment like
                 Breakout. Only used for evaluation.
-            eval_uncertainty (bool): whether to evluate uncertainty after training
+            eval_uncertainty (bool): whether to evluate uncertainty after training.
+            hold_out_dataset (str): name of hold out dataset.
+            train_classes (list[int]): class indices to use for training.
+            hold_out_classes (list[int]): class indices to exclude from training.
```
Comment (emailweixu) [inline alf/algorithms/config.py]: There are four arguments commented here. But there are only 3 added at line 41.

File: alf/algorithms/config.py [hunk 7]
```diff
@@ -38,6 +38,9 @@ def __init__(self,
                  eval_interval=10,
                  epsilon_greedy=0.,
                  eval_uncertainty=False,
+                 hold_out_dataset=None,
+                 train_classes=None,
+                 hold_out_classes=None,
```
Comment (emailweixu) [inline alf/algorithms/config.py]: need to assign these values to self

File: alf/algorithms/config.py (hunk 6 above)
Comment (emailweixu) [inline alf/algorithms/config.py]: Are these configs actually being used somewhere in the code? I couldn't find.

File: alf/algorithms/functional_particle_vi_algorithm.py [hunk 8]
```diff
@@ -485,6 +520,43 @@ def _regression_vote(self, output, target):
         total_loss = regression_loss(output, target)
         return loss, total_loss
 
+    def eval_uncertainty(self):
+        """
+        Function to evaluate the epistemic uncertainty of the ensemble.
```
Comment (emailweixu) [inline alf/algorithms/functional_particle_vi_algorithm.py]: Put this line immediately after `"""`. Add a blank line after it. The rest of the comment need to be aligned with the first character of `"""`. See https://alf.readthedocs.io/en/latest/notes/howto_docstring.html for reference

File: alf/algorithms/functional_particle_vi_algorithm.py [hunk 9]
```diff
@@ -485,6 +520,43 @@ def _regression_vote(self, output, target):
...
+            This method computes the following metrics:
+        
+            * AUROC (AUC) evaluates the separability of model predictions with
+                respect to the training data and a prespecified outlier dataset.
```
Comment (emailweixu) [inline alf/algorithms/functional_particle_vi_algorithm.py]: should align with the first char of AUROC

File: alf/utils/sl_utils.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        auc_score = roc_auc_score(y_true, y_score)
+    except NameError:
+        absl.logging.info('roc_auc_score function not defined')
+        auc_score = 0.5
+    return auc_score
+
+
+def predict_dataset(model, testset):
+    """
+    Computes predictions for an input dataset.
```
Comment (emailweixu) [inline alf/utils/sl_utils.py]: Put this immediately after `"""` and add a blank line after it. The docstring should look like the following. Please also fix other cases too. See https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html for examples.

File: alf/utils/sl_utils.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+def predict_dataset(model, testset):
+    """
+    Computes predictions for an input dataset. 
+    Args: 
+        model (Callable): model with which to compute predictions.
+        testset (torch.utils.data.DataLoader): dataset for which to compute
+            predictions.
+
+    Returns:
+        model_outputs (torch.tensor): a tensor of shape [N, S, D] where
```
Comment (emailweixu) [inline alf/utils/sl_utils.py:112]: This functions return a tuple of Tensors, both need to be documented.

File: alf/algorithms/functional_particle_vi_algorithm.py [hunk 12]
```diff
@@ -416,12 +452,15 @@ def _neglogprob(self, inputs, params):
...
+                                                num_particles)
 
         return self._loss_func(output, target)
 
-    def evaluate(self, num_particles=None):
-        """Evaluate on the fixed ensemble. """
+    def evaluate(self):
+        """
+        Evaluatation of the ParVI ensemble on a test dataset
+        """
```
Comment (emailweixu) [inline alf/algorithms/functional_particle_vi_algorithm.py]: make it one line.

File: alf/algorithms/config.py (hunk 7 above)
Comment (emailweixu) [inline alf/algorithms/config.py]: I feel that these options are too algorithm specific. Instead of putting here, they should be the arguments of the related algorithm

Comment (neale) [reply alf/algorithms/config.py]: The easiest entry point to handle datasets is within the SL_trainer class in the policy_trainer file.
One compromise I see is to load the full datasets within the trainer, but handle what classes are used for training/testing within the algorithm. In that way, these options don't need to be part of config. 
But I think it would be odd to handle all dataset creation within the algorithm.

Comment (emailweixu) [reply alf/algorithms/config.py]: The problem is that for supervised learning, the form of data can be quite different. I made a refactoring so that the algorithm will have an argument `data_creator` for creating data see PR #840.

Comment (neale) [reply alf/algorithms/config.py]: I see, this is good. I'll incorporate this data creator.

File: alf/utils/sl_utils.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    pass
+
+
+def classification_loss(output, target):
+    """
+    Computes the cross entropy loss with respect to a batch of predictions and
+        targets.
+    
+    Args:
+        output (Tensor): predictions of shape ``[B, D]`` or ``[N, B, D]``
```
Comment (runjerry) [inline alf/utils/sl_utils.py]: [N, B, D] -> [B, N, D]

File: alf/utils/sl_utils.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+
+def classification_loss(output, target):
+    """
+    Computes the cross entropy loss with respect to a batch of predictions and
+        targets.
+    
+    Args:
+        output (Tensor): predictions of shape ``[B, D]`` or ``[N, B, D]`` 
+        target (Tensor): targets of shape ``[B, 1]`` or ``[B, N, 1]``
```
Comment (runjerry) [inline alf/utils/sl_utils.py]: ``[B]``, ``[B, 1]``, ``[B, N]``, or ``[B, N, 1]``

File: alf/utils/sl_utils.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+            accuracy.
+    """
+
+    if output.ndim == 2:
+        output = output.reshape(output.shape[0], target.shape[1], -1)
+    pred = output.max(-1)[1]
+    target = target.squeeze(-1)
+    acc = pred.eq(target).float().mean(0)
+    avg_acc = acc.mean()
+    if output.dim == 3:
```
Comment (runjerry) [inline alf/utils/sl_utils.py]: if output.ndim == 3:

File: alf/utils/sl_utils.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    return LossInfo(loss=loss, extra=avg_acc)
+
+
+def regression_loss(output, target):
+    """
+    Computes the MSE loss with respect to a batch of predictions and
+        targets.
+    
+    Args:
+        output (Tensor): predictions of shape ``[B, 1]`` or ``[N, B, 1]``
```
Comment (runjerry) [inline alf/utils/sl_utils.py]: [N, B, 1] -> [B, N, 1]

**Review by runjerry — CHANGES_REQUESTED** [review]

File: alf/algorithms/config.py (hunk 6 above)
Comment (neale) [reply alf/algorithms/config.py]: I removed these arguments when adding in the new data loading scheme introduced in the latest PR.

File: alf/algorithms/functional_particle_vi_algorithm.py (hunk 8 above)
Comment (neale) [reply alf/algorithms/functional_particle_vi_algorithm.py]: I think I fixed the docstring issues in the files that I committed. Let me know if I missed one.

File: alf/utils/sl_utils.py (hunk 13 above)
Comment (neale) [reply alf/utils/sl_utils.py]: Fixed these docstring issues that you have outlined.

**Review by runjerry — APPROVED** [review]
